File: examples/python/hammerhead_scheduler/hammerhead_scheduler/hammerhead_scheduler.py
import struct
import sys
import time

import zmq


class Scheduler:
    def __init__(self, endpoint):
        self.context = zmq.Context()
        self.socket = self.context.socket(zmq.ROUTER)
        self.socket.connect(endpoint)
        print(f"Subscribing to {endpoint}")
        self.last_frame_id = 0
        self.initialized = False

    def loop(self):
        # Wait for a scheduler request from hammerhead indicating that a frame is ready.
        identity_msg = self.socket.recv()
        request_msg = self.socket.recv()

        if len(request_msg) != 8:
            print("The scheduler request doesn't seem to be the right size.")
            return

        # Deserialize the request to get the frame id
        # https://docs.python.org/3/library/struct.html#format-characters
        frame_id = struct.unpack('Q', request_msg)[0]
        print(f"\rGot a scheduler request for frame {frame_id}", end='', flush=True)
        if self.initialized and frame_id != self.last_frame_id + 1:
            print(f"\nIt looks like we might be dropping frames\n"
                  f"Current frame_id = {frame_id}\n"
                  f"Last    frame_id = {self.last_frame_id}")

        self.initialized = True
        self.last_frame_id = frame_id

        # Halt hammerhead execution before telling it to continue
        time.sleep(0.5)
        reply = b''  # Intentionally empty message
        self.socket.send(identity_msg, zmq.SNDMORE)
        self.socket.send(reply)


DEFAULT_IP = "127.0.0.1"


def print_usage():
    print("You should specify the IP address of the device running hammerhead:\n\n"
          "     python hammerhead_scheduler.py IP\n\n"
          "e.g. python hammerhead_scheduler.py 192.168.1.9\n\n"
          "In the meantime, we are going to assume that you are running this on the same device as hammerhead,\n"
          f"that is, we assume that you specified {DEFAULT_IP}"
          "\n----------------------------------------"
          )


def main():
    if len(sys.argv) == 1:
        print_usage()

    ip = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_IP
    endpoint = f"tcp://{ip}:9814"  # Corresponds to WAIT_TOPIC
    scheduler = Scheduler(endpoint)

    # Infinite loop that runs
    try:
        while True:
            scheduler.loop()
    except KeyboardInterrupt:
        print("\nExiting...")


if __name__ == "__main__":
    main()

File: examples/cpp/hammerhead_scheduler/src/hammerhead_scheduler.cpp
#include <chrono>
#include <iostream>
#include <memory>
#include <nodar/zmq/topic_ports.hpp>
#include <thread>
#include <zmq.hpp>

class Scheduler {
public:
    Scheduler(const std::string &endpoint) : context(1), socket(context, ZMQ_ROUTER) {
        socket.connect(endpoint);
        std::cout << "Subscribing to " << endpoint << std::endl;
    }

    void loop() {
        // Wait for a scheduler request from hammerhead indicating that a frame is ready.
        zmq::message_t identity_msg;
        const auto identity = socket.recv(identity_msg, zmq::recv_flags::none);
        if (not identity) {
            std::cerr << "There was an error waiting for the scheduler request's identity message" << std::endl;
            return;
        }
        ::zmq::message_t request_msg;
        const auto request = socket.recv(request_msg);
        if (not request) {
            std::cerr << "There was an error waiting for the scheduler request" << std::endl;
            return;
        }
        if (request_msg.size() != sizeof(uint64_t)) {
            std::cerr << "The scheduler request doesn't seem to be the right size." << std::endl;
            return;
        }

        // Deserialize the request to get the frame id
        uint64_t frame_id;
        std::memcpy(&frame_id, request_msg.data(), sizeof(frame_id));
        std::cout << "\rGot a scheduler request for frame " << frame_id << std::flush;
        if (initialized and frame_id != last_frame_id + 1) {
            std::cerr << "\nIt looks like we might be dropping frames\n"
                      << "Current frame_id = " << frame_id << "\n"
                      << "Last    frame_id = " << last_frame_id << std::endl;
        }
        initialized = true;
        last_frame_id = frame_id;

        // Halt hammerhead execution before telling it to continue
        std::this_thread::sleep_for(std::chrono::milliseconds(300));
        zmq::message_t reply(0);  // Intentionally empty message
        socket.send(identity_msg, zmq::send_flags::sndmore);
        socket.send(reply, zmq::send_flags::none);
    }

private:
    zmq::context_t context;
    zmq::socket_t socket;
    uint64_t last_frame_id{0};
    bool initialized = false;
};

constexpr auto DEFAULT_IP = "127.0.0.1";

void printUsage() {
    std::cout << "You should specify the IP address of the device running hammerhead:\n\n"
                 "     ./hammerhead_scheduler IP\n\n"
                 "e.g. ./hammerhead_scheduler 10.10.1.10\n\n"
                 "In the meantime, we assume that you are running this on the device running Hammerhead,\n"
                 "that is, we assume that you specified "
              << DEFAULT_IP << "\n----------------------------------------" << std::endl;
}

int main(int argc, char **argv) {
    if (argc == 1) {
        printUsage();
    }
    const auto ip = argc > 1 ? argv[1] : DEFAULT_IP;
    const auto endpoint = std::string("tcp://") + ip + ":" + std::to_string(nodar::zmq::WAIT_TOPIC.port);
    Scheduler scheduler(endpoint);

    // Infinite loop that runs
    while (true) {
        scheduler.loop();
    }
    return 0;
}